### PR TITLE
[AMDGPU] Introduce iglp_opt(2): Generalized exp/mfma interleaving for select kernels

### PR DESCRIPTION
**Pull request by jrbyrnes**
This is a redesign of https://github.com/llvm/llvm-project/pull/80370 . The implementation is a bit more organized and less repetitive.

This implements the basic pipelining structure of exp/mfma interleaving for better extensibility. While it does have improved extensibility, there are controls which only enable it for DAGs with certain characteristics (matching the DAGs it has been designed against).

### CODE REVIEW

**Comment by jrbyrnes** [comment]
Delete support for tiny kernel since the kernel itself will not be released until next iteration.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.h [hunk 1]
```diff
@@ -11,10 +11,17 @@
...
 #include "llvm/CodeGen/ScheduleDAGMutation.h"
 #include <memory>
+#include <vector>
 
 namespace llvm {
 
-std::unique_ptr<ScheduleDAGMutation> createIGroupLPDAGMutation(bool IsReentry);
+// Components of the mask that determines which instruction types may be may be
+// classified into a SchedGroup.
+enum class IGLPPhase { Initial = 0u, PreRAReentry = 1u << 0, PostRA = 1u << 1 };
```
[the remark this answers is not in the document]
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.h]: Doesn't need to be mask

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 2]
```diff
@@ -902,6 +904,921 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+  /// instruction
+  class IsPipeExp final : public InstructionRule {
+  public:
+    bool apply(const SUnit *SU, const ArrayRef<SUnit *> Collection,
+               SmallVectorImpl<SchedGroup> &SyncPipe) override {
+
+      auto DAG = SyncPipe[0].DAG;
+      auto TII = SyncPipe[0].TII;
+
+      if (Cache->empty()) {
```
[the remark this answers is not in the document]
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:947]: This rule is only currently used in MFMAExpInterleaveOpt which is only applied if analyzeDAG finds that the DAG meets certain criteria. One of these criteria is that there exists at least one dependency chain wherein a V_EXP is a transitive predecessor of a V_MFMA. 

Perhaps it would be a better design to have analyzeDAG produce a "IGLPDAGInfo" instance which controls behavior of the pipeline and rules. Either way, in the current implementation we won't run into the problem where we are consistently looking for a non-existant MFMA.

Parsing bottom up we find the last MFMA, which falls into the third stage of the CK pipeline (see below).

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 3]
```diff
@@ -902,6 +904,921 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+          }));
+
+      return Reaches;
+    }
+    IsPipeExp(const SIInstrInfo *TII, unsigned SGID, bool NeedsCache = false)
+        : InstructionRule(TII, SGID, NeedsCache) {}
+  };
+
+  /// Whether or not the instruction enables the exact MFMA that is the \p
+  /// Number th MFMA in the chain starting with \p ChainSeed
```
[the remark this answers is not in the document]
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:970]: There are dependencies among the (stage 3, see below) MFMAs themselves; there are running sums of MFMAs. 

e.g. 
	MFMA[0,0] = MFMA A00, B00, C
	MFMA[0,1] = MFMA A01, B01, MFMA[0,0]
	...

	MFMA[1,0] = MFMA A10, B10, C
	MFMA[1,1] = MFMA A11, B11, MFMA[1,0]
	...

	...


MFMAChainSeed would be those MFMAs that do not have an MFMA predecessor (MFMA[i,0] in this example). The chain length would be the max value of j + 1 in MFMA[0,j].

As another example, if there are 16 MFMAs and none of them depend on each other, we will have 16 MFMAChainSeeds.

I think if there was documentation on this pattern it would exist somewhere in CK materials regarding the implementation of the fused_attention operation.

In terms of interleaving, this means that we must put MFMA[0,0] into an earlier MFMA slot than MFMA[0,1]. We must also be careful to put MFMA[0,1]'s V_EXP predecessors in the correct slots as well. Ideally, the greedy PipelineSolver would be able to figure all this stuff out, but it can't.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 4]
```diff
@@ -902,6 +904,921 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+
+        auto I = DAG->SUnits.begin();
+        auto E = DAG->SUnits.end();
+        for (; I != E; I++) {
+          if (!FoundTrans) {
+            if (TII->isTRANS(I->getInstr()->getOpcode()))
+              FoundTrans = true;
+            continue;
+          } else {
+            if (TII->isMFMAorWMMA(*I->getInstr())) {
```
[the remark this answers is not in the document]
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp]: The basic stages of the CK pipeline for this fused attention kernels are:
	1. A series of MFMA chains.
	2. A series of V_EXP on the outputs of 1.
	3. Another series of MFMA chains on the ouputs of 2.

(see the IR in corresponding *.ll)

Since each EXP is dependent upon all MFMAs in stage 1 we cannot interleave those. However, each MFMA in stage 3 is only dependent on a subset of V_EXPs in stage 2, so we can interleave those so long as we carefully pick which EXP / MFMA go into a which "interleaving slot" (i.e. SchedGroup in the pipeline). In other words, the only MFMAs relevant to interleaving for the subset of kernels to which this is applied are the MFMAs that occur after the first V_EXP.

Some other contextual info:

In PreRA we can analyze the dependency relationships to understand exactly which V_EXPs are needed for which stage 3 MFMAs, thus we use EnablesNthMFMAInChain to match an exp which produces an exact MFMA (and IsExactMFMA to match the exact MFMA).

In PostRA, the dependency relationships are muddied via PhysReg so we cannot do the exact analysis. However, we assume that the PreRA mutations have been successful, thus the V_EXP and V_MFMA are already in the relative order we want. We just need to preserve that relative ordering; hence EnablesNthMFMA

**Comment by jrbyrnes** [comment]
Don't allow multiple entries into initIGLPOpt

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kerbowa) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:947]: Can it be assumed based on the ordering of SUnits that it will be topo sorted and therefore will be added in the correct order?

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 5]
```diff
@@ -902,6 +904,926 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+              Cache->push_back(&*I);
+              break;
+            }
+            ++Counter;
+          }
+          if (!FoundTrans && TII->isTRANS(I->getInstr()->getOpcode()))
+            FoundTrans = true;
+        }
+      }
+      if (Cache->empty())
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1000]: Can move this to the block above

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 6]
```diff
@@ -902,6 +904,926 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+      auto DAG = SyncPipe[0].DAG;
+
+      if (Cache->empty()) {
+        auto TII = SyncPipe[0].TII;
+        SmallVector<SUnit *, 8> Worklist;
+
+        auto I = DAG->SUnits.begin();
+        auto E = DAG->SUnits.end();
+        for (; I != E; I++) {
+          if (FoundTrans && TII->isMFMAorWMMA(*I->getInstr())) {
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:989]: Like in the rule above, Is there a reason to only start counting MFMA after the first "Trans" instruction is seen?

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 7]
```diff
@@ -902,6 +904,926 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+
+      return DAG->IsReachable((*Cache)[0], const_cast<SUnit *>(SU));
+    }
+
+    EnablesNthMFMA(unsigned Number, const SIInstrInfo *TII, unsigned SGID,
+                   bool NeedsCache = false)
+        : InstructionRule(TII, SGID, NeedsCache), Number(Number) {}
+  };
+
+  /// Whether or not the instruction enables  a transitive predecessor of the
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1011]: Extra space in the comment.

Does "enables" just mean is a predecessor of, so this rule is checking whether SU is a predecessor of the Nth MFMA in a the MFMA chain?

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 8]
```diff
@@ -902,6 +904,926 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+                   bool HasIntermediary = false, bool NeedsCache = false)
+        : InstructionRule(TII, SGID, NeedsCache), Size(Size),
+          HasIntermediary(HasIntermediary) {}
+  };
+
+  /// Whether or not the instruction has greater than or equal to \p Size
+  /// immediate successors. If \p HasIntermediary is true, this tests also
+  /// whether all successors of the SUnit have greater than or equal to \p Size
+  /// successors.
+  class GreaterThanNSuccs final : public InstructionRule {
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1104]: Probably should incorporate "or equal" into the name of the function.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 9]
```diff
@@ -902,6 +904,926 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+        return false;
+
+      for (auto &PipeSG : SyncPipe) {
+        if ((unsigned)PipeSG.getSGID() == SGID - Distance)
+          OtherGroup = &PipeSG;
+      }
+
+      if (!OtherGroup)
+        return false;
+      if (!OtherGroup->Collection.size())
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1225]: Why is this true if there are no instructions in the other group.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 10]
```diff
@@ -902,6 +904,926 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+
+      return false;
+    }
+    IsReachableFromPrevNthGroup(unsigned Distance, const SIInstrInfo *TII,
+                                unsigned SGID, bool NeedsCache = false)
+        : InstructionRule(TII, SGID, NeedsCache), Distance(Distance) {}
+  };
+
+  /// Whether or not the instruction occurs after the SU with NodeNUm \p Number
+  class OccursAfterDSRead final : public InstructionRule {
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1242]: Is this specific to DSRead?

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 11]
```diff
@@ -902,6 +904,926 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+  if (TempCvt.has_value()) {
+    for (auto &SuccSU : MFMAPipeSUs) {
+      if (DAG->IsReachable(SuccSU, *TempCvt)) {
+        HasCvt = true;
+        break;
+      }
+    }
+  }
+
+  MFMAChains = 0;
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1467]: Would be nice to eventually refactor things so that rules can access shared data for the strategy directly, e.g. if you could just feed in the MFMA chains to the rules directly.

Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1467]: Yes it would be nice if we could reduce the amount of DAG processing -- potentially by producing some fields / a class in the Strategy which the rules could access.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp (hunk 9 above)
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1225]: Changed since it is NFC. I believe this pattern emerged during design of iglp_opt(1) -- there was an issue where PhysReg deps were causing an inability to assign to one group, which resulted in not assigning to another (!OtherGroup->Collection.size()) and caused the whole pipeline to break. However, I believe that issue was resolved via redesign of the rules.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp (hunk 7 above)
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1011]: I've used "enabled" in a few places -- it means "is a transitive predecessor of". I've fixed this particular comment which was just wrong somehow.

IsATransitivePredOf felt a bit wordy.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp (hunk 6 above)
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:989]: In the relevant kernels, there are a group of MFMAs which are all transitive predecessors of every V_EXP and there is an MFMAs group wherein each MFMA is a transitive sucessors of a subset of the V_EXPs. The first group of MFMAs must occur before the first V_EXP and the second group of MFMAs must occur after the first V_EXP. There is no interleaving opportunity between the first group of MFMAs and the V_EXPs, since all the V_EXPs are dependent upon all the V_MFMAs. However, we can interleave within the second group.

Thus, we are only concerned (in this iteration) with the MFMAs occuring after the V_EXPs. We could count all MFMAs, but then we will be looking for the (MFMAPredCount + Number)th MFMA, wherein MFMAPredCount is constant. Only considering MFMAs after EXP captures what we need, and saves us from the redundant MFMAPredCount.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:970]: Oh. Yeah, the comment is just wrong... Thanks.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 12]
```diff
@@ -902,6 +906,923 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+
+  HasChainBetweenCvt =
+      std::find_if((*TempExp)->Succs.begin(), (*TempExp)->Succs.end(),
+                   [&isCvt](SDep &Succ) {
+                     return isCvt(Succ.getSUnit()->getInstr()->getOpcode());
+                   }) == (*TempExp)->Succs.end();
+
+  // Count the number of MFMAs that are reached by an EXP
+  MFMAPipeCount = 0;
+  for (auto &SuccSU : MFMAPipeCands) {
```
[the remark this answers is not in the document]
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1425]: Yes, MFMAPipeCnads are all MFMA, and MFMAPipeSUs are the MFMA which have V_EXP (transitive) predcessor.

So yes, We can more efficiently determine MFMAPipeCount in this way.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 13]
```diff
@@ -902,6 +906,923 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
...
+      if (DAG->IsReachable(SuccSU, *TempCvt)) {
+        HasCvt = true;
+        break;
+      }
+    }
+  }
+
+  MFMAChains = 0;
+  for (auto &MFMAPipeSU : MFMAPipeSUs) {
+    if (MFMAChainSeeds.size() &&
```
[the remark this answers is not in the document]
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1465]: They may be different in subsequent phases.

**Review by kerbowa — APPROVED** [review]
LGTM

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 14]
```diff
@@ -1165,6 +1168,18 @@ class MFMAExpInterleaveOpt final : public IGLPStrategy {
         : InstructionRule(TII, SGID, NeedsCache) {}
   };
 
+  // Whether or not the instruction is a V_ADD_F32 instruction.
+  class IsPipeAdd final : public InstructionRule {
+  private:
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1173]: Remove

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 15]
```diff
@@ -1726,6 +1749,8 @@ bool MFMAExpInterleaveOpt::applyIGLPStrategy(
   // In loop MFMAs
   auto MFMAInLoop = MFMAPipeCount - (MFMAEnablement * 2);
   auto MFMALoopCount = MFMAInLoop / MFMARatio;
+  auto VALUOps =
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1752]: Maybe just double-check whether any of these divs need a check for div-by-zero.

**Review by kerbowa — APPROVED** [review]
LGTM with NIT

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp (hunk 15 above)
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp]: Added a couple checks to be safe, but we should be good

**Comment by jrbyrnes** [comment]
Pushing to use GUI for merge.
